runtime/src/accounts_db.rs

```diff
@@ -2209,67 +2251,186 @@ impl AccountsDb {
+    // and at the tip of its fork
+    pub fn load_slow_with_fixed_root(
         &self,
         ancestors: &Ancestors,
         pubkey: &Pubkey,
-        max_root: Option<Slot>,
     ) -> Option<(Account, Slot)> {
-        let (slot, store_id, offset) = {
-            let (lock, index) = self.accounts_index.get(pubkey, Some(ancestors), max_root)?;
-            let slot_list = lock.slot_list();
-            let (
-                slot,
-                AccountInfo {
-                    store_id, offset, ..
-                },
-            ) = slot_list[index];
-            (slot, store_id, offset)
-            // `lock` released here
-        };
+        self.do_load(ancestors, pubkey, None, true)
+    }
 
-        //TODO: thread this as a ref
-        self.get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset)
-            .get_loaded_account()
-            .map(|loaded_account| (loaded_account.account(), slot))
+    fn accounts_index_get(
+        &self,
+        ancestors: &Ancestors,
+        pubkey: &Pubkey,
+        max_root: Option<Slot>,
+    ) -> Option<(Slot, AppendVecId, usize)> {
+        let (lock, index) = self.accounts_index.get(pubkey, Some(ancestors), max_root)?;
+        let slot_list = lock.slot_list();
+        let (
+            slot,
+            AccountInfo {
+                store_id, offset, ..
+            },
+        ) = slot_list[index];
+        Some((slot, store_id, offset))
     }
 
-    pub fn load_account_hash(&self, ancestors: &Ancestors, pubkey: &Pubkey) -> Hash {
-        let (slot, store_id, offset) = {
-            let (lock, index) = self
-                .accounts_index
-                .get(pubkey, Some(ancestors), None)
-                .unwrap();
-            let slot_list = lock.slot_list();
-            let (
-                slot,
-                AccountInfo {
-                    store_id, offset, ..
-                },
-            ) = slot_list[index];
-            (slot, store_id, offset)
-            // lock released here
-        };
+    fn get_account_accessor_with_retry<'a>(
+        &'a self,
+        ancestors: &'a Ancestors,
+        pubkey: &'a Pubkey,
+        max_root: Option<Slot>,
+        is_root_fixed: bool,
+    ) -> Option<(LoadedAccountAccessor<'a>, Slot)> {
+        let (mut slot, mut store_id, mut offset) =
+            self.accounts_index_get(ancestors, pubkey, max_root)?;
 
-        self.get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset)
-            .get_loaded_account()
-            .map(|loaded_account| *loaded_account.loaded_hash())
-            .unwrap()
+        #[cfg(test)]
+        {
+            // Give some time for cache flushing to occur here for unit tests
+            sleep(Duration::from_millis(self.load_delay));
+        }
+
+        let mut account_accessor =
+            self.get_account_accessor_from_cache_or_storage(slot, pubkey, store_id, offset);
```

Comment:
Maybe, we can simplify this crutial looping code a bit by unifying `get_account_accessor_from_cache_or_storage` below (https://github.com/solana-labs/solana/pull/15454/files#r590768524), by moving this call to the top inside the `loop`?